### ISSUE TITLE
Set data type

### ISSUE THREAD

**Issue by toastal** [tweag/nickel#1735]
**Is your feature request related to a problem? Please describe.**
For collection types, Nickel supports: records, dictionary, arrays, but one common structure missing is `Set`, an unordered collection like an array where all values must be unique. Uniqueness can prevent certain kinds of errors where the consumer should need to consider duplicate values, but also comes with a grab bag of handy combinators like `union`, `intersect`, `difference`, `merge`, etc.

**Describe the solution you'd like**
Add a new data type: `Set T` + `stdlib` module `std-set`

**Describe alternatives you've considered**
Making a bunch of contracts around `Array T`.

**Additional context**
This is a common data structure in most languages. It’s been a long personal pain point with its absence in JSON, however alternatives like EDN do support it. Being able to assert a JSON Array as a Set would be very useful.

**Comment by yannham** [comment]
I wonder, do you want a builtin type `Set` that would be natively handled by Nickel, or some kind of standard library module that implements `Set` using more basic constructions? Or, maybe, you don't care and just want `Set` :slightly_smiling_face:

**Comment by suimong** [comment]
Weighing in here as someone who also encountered use cases for `Set`, I believe a native implementation would be nicer, though the only reason I could think of is performance.😂

**Comment by thufschmitt** [comment]
Agreed that a `Set` could be useful.

I had fun with the encoding below, which has the advantage of playing nicely with the merge system (because `a & b` is the set-theoretic merge), but doesn't translate nicely to JSON at all:

```nickel
let Set = {
  Set | doc "Contract for a set" = fun a => { _ : a },
  hashFun : Dyn -> String = fun x => x |> std.serialize 'Json |> std.hash 'Sha256,
  add
    | doc "Add an element to a set"
    : forall a. a -> { _ : a } -> { _ : a }
    = fun elt set =>
      let key : String = hashFun (elt | Dyn) in
      std.record.update key elt set,

  add_all
    : forall a. Array a -> { _ : a } -> { _ : a }
    = fun elts set =>
      std.array.fold_left
        (fun x y => add y x)
        set
        elts,

  from_array
    | doc m%"
      Create a set from all the elements in the given array.
    "%
    : forall a. Array a -> { _ : a }
    = fun elts => add_all elts {},
}
in
# Just for the example
(
  Set.from_array [1, 3, 5, 8, 9, 143]
  & (Set.add_all [1, 2, 3] {})
  & (Set.add_all [2, 2, 3] {})
) | Set.Set Number

```

**Comment by yannham** [comment]
Coming back to this idea, there's a another question to answer: how would sets be serialized? One obvious answer is as arrays, but then we need to make the order deterministic, or this is going to be mess. Maybe as an array where elements are ordered by hash values is reasonable? Another possibility is to keep insertion order, but it's not so well defined if we start to implement unions, intersections or whatnot.

**Comment by yannham** [comment]
I agree that it's an implementation detail, but I think it has to be at the very least deterministic, otherwise we can imagine having some tests comparing expected outputs randomly failing just because Nickel chose a different order this time.

**Comment by suimong** [comment]
Oh yes, absolutely. What I mean is since this is an implementation detail, thus constraints on the implementation side, in this case determinism of order and maybe other factors, should be the sole concerns while making design decisions, and using hash-ordered array seems like a natural choice. Insertion order cannot be meaningfully defined in some cases, eg converting an array to a set using parallel algorithm.

On the other hand, I'm starting to wonder why you think "serialized order" is important. For the sake of argument, there are certain assumptions to be made: I assume "serialize a set" means to represent a set in a nickel file, and as we are unlikely to add dedicated syntax to represent a set, we will instead use array (plus static typing or contract) to "hold" set values. I'm also assuming that you are worried about non-deterministic order to complicate implementations of set operations. Two questions: 1. shouldn't set operations be done on "deserialized" sets? I'm assuming we perform "deserialization" by hashing, so does order really matter? 2. Whatever order we choose to serialize a set to an array, user can always manually change the ordering. After all it's just a file and they can do whatever they want. In this sense, deterministic order just cannot be guaranteed.

**Comment by yannham** [comment]
Yes, by serialization I mean what is the result of exporting a value that is a set to JSON. If we represent a set as a Nickel array then I guess the questions is moot, but another solution could be to have a new bespoke, primitive set datatype.

**Comment by yannham** [comment]
> how nickel array elements and record fields are sorted, which I believe is alphanumeric order.

Record fields are sorted alphabetically, but arrays have an inherent order, that is of course preserved. Nickel doesn't re-order arrays behind your back 🙂 

> Actually I'm not sure if this is true, does introducing an internal set type implies new syntax? Can we instead use eg array + static typing to achieve the same?

You can have a new primitive type without syntax. One way would be to provide constructors and destructors into existing primitive types: for example `std.set.from_array`, `std.set.to_array`, and operations in `std.set.***`. The only difference with using arrays as backing repr, beside having to hard-code set in the interpreter, is precisely to decide what to do when serializing them. Also pretty-printing is a bit harder without syntax, but you can always represent it as an array and replace set objects by `std.set.from_array <array repr of the set>`, I suppose.
